pkg/segment/query/iqr/intermediateQueryResult.go

```diff
@@ -716,17 +718,162 @@ func (iqr *IQR) AsResult() (*structs.PipeSearchResponseOuter, error) {
-		Qtype:              "logs-query", // TODO: handle stats queries
-		CanScrollMore:      false,
-		ColumnsOrder:       toputils.GetSortedStringKeys(records),
+	var response *structs.PipeSearchResponseOuter
+
+	switch iqr.qType {
+	case structs.RRCCmd:
+		response = &structs.PipeSearchResponseOuter{
+			Hits: structs.PipeSearchResponse{
+				TotalMatched: iqr.NumberOfRecords(),
+				Hits:         recordsAsAny,
+			},
+			AllPossibleColumns: toputils.GetKeysOfMap(records),
+			Errors:             nil,
+			Qtype:              iqr.qType.String(),
+			CanScrollMore:      false,
+			ColumnsOrder:       toputils.GetSortedStringKeys(records),
+		}
+	case structs.GroupByCmd:
+		bucketHolderArr, aggGroupByCols, measureFuncs, bucketCount, err := iqr.getFinalResultForGroupBy()
+		if err != nil {
+			return nil, toputils.TeeErrorf("IQR.AsResult: error getting final result for GroupBy: %v", err)
+		}
+
+		response = &structs.PipeSearchResponseOuter{
+			Hits: structs.PipeSearchResponse{
+				TotalMatched: 0, // TODO: get the total matched records
+				Hits:         nil,
+			},
+			AllPossibleColumns: toputils.GetKeysOfMap(records),
+			Errors:             nil,
+			Qtype:              iqr.qType.String(),
+			CanScrollMore:      false,
+			ColumnsOrder:       toputils.GetSortedStringKeys(records),
+			BucketCount:        bucketCount,
+			MeasureFunctions:   measureFuncs,
+			MeasureResults:     bucketHolderArr,
+			GroupByCols:        aggGroupByCols,
+			// TODO: add IsTimechart flag
+		}
+	case structs.SegmentStatsCmd: // TODO: implement
+		panic("IQR.AsResult: SegmentStatsCmd not implemented")
+	default:
+		return nil, fmt.Errorf("IQR.AsResult: unexpected query type %v", iqr.qType)
 	}
 
 	return response, nil
 }
+
+func (iqr *IQR) AppendGroupByResults(bucketHolderArr []*structs.BucketHolder, measureFuncs []string, aggGroupByCols []string, bucketCount int) error {
+	if err := iqr.validate(); err != nil {
+		log.Errorf("IQR.AppendGroupByResults: validation failed: %v", err)
+		return err
+	}
+
+	if iqr.qType != structs.GroupByCmd {
+		return fmt.Errorf("IQR.AppendGroupByResults: not a GroupBy query. queryType=%v", iqr.qType)
+	}
+
+	iqr.mode = withoutRRCs
+
+	knownValues := make(map[string][]utils.CValueEnclosure)
+
+	for _, aggGroupByCol := range aggGroupByCols {
+		knownValues[aggGroupByCol] = make([]utils.CValueEnclosure, bucketCount)
+	}
+	for _, retMFun := range measureFuncs {
+		knownValues[retMFun] = make([]utils.CValueEnclosure, bucketCount)
+	}
+
+	conversionErrors := make([]string, utils.MAX_SIMILAR_ERRORS_TO_LOG)
+	errIndex := 0
+
+	for i, bucketHolder := range bucketHolderArr {
+		for idx, aggGroupByCol := range aggGroupByCols {
+			colValue := bucketHolder.GroupByValues[idx]
+			err := knownValues[aggGroupByCol][i].ConvertValue(colValue)
+			if err != nil && errIndex < utils.MAX_SIMILAR_ERRORS_TO_LOG {
+				conversionErrors[i] = fmt.Sprintf("BucketHolderIndex=%v, groupByCol=%v, ColumnValue=%v", i, aggGroupByCol, colValue)
```

Comment:
Is `conversionErrors[i] = ...` supposed to be `conversionErrors[errIndex] = ...`?

Same question below

Reply:
Ahh Yes, my bad.

Reply:
Fixed in the PR: #1774